components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/internal/impl/UserSessionManagementServiceImpl.java

```diff
@@ -605,10 +609,111 @@ private List<UserSession> getActiveSessionList(List<String> sessionIdList, Strin
                 }
             }
         }
-
         return sessionsList;
     }
 
+    private boolean isEffectiveSession(SessionContext sessionContext, UserSession userSession) {
+
+        try {
+            String sessionTenantDomain = null;
+            if (sessionContext.getProperties() != null &&
+                    sessionContext.getProperties().get(FrameworkUtils.TENANT_DOMAIN) instanceof String) {
+                sessionTenantDomain = (String) sessionContext.getProperties().get(FrameworkUtils.TENANT_DOMAIN);
+            }
+            if (StringUtils.isEmpty(sessionTenantDomain)) {
+                return true;
+            }
+            String loginTenantDomain = FrameworkUtils.getLoginTenantDomainFromContext();
+            boolean isOrganization = OrganizationManagementUtil.isOrganization(loginTenantDomain);
+            if (StringUtils.equals(sessionTenantDomain, loginTenantDomain)) {
+                // This block handles the scenario where session tenant domain and login tenant domain are equal.
+                if (isOrganization & areAllFragmentAppsInUserSession(userSession)) {
+                   /*
+                   When a sub org user authenticates using a shared application, sessions are created for shared app in
+                   sub organization and parent app in primary organization. In this case, the session created in primary
+                   organization is the effective session. Hence, this session is ignored for the session list.
+                    */
+                    return false;
+                } else {
```

Comment:
Do we need to have the else block here?

Reply:
We can get rid of this else block and return true after all if else blocks. But kept it this way to separate every case clearly.